src/bun.js/node/node_fs_watcher.zig

```diff
@@ -540,241 +562,202 @@ pub const FSWatcher = struct {
         }
-        pub fn emitError(this: *JSObject, err: string) void {
-            if (this.closed) return;
-            defer this.close(true);
-
-            if (this.globalThis) |globalThis| {
-                if (this.js_this != .zero) {
-                    if (JSObject.listenerGetCached(this.js_this)) |listener| {
-                        var args = [_]JSC.JSValue{
-                            JSC.ZigString.static("error").toValue(globalThis),
-                            JSC.ZigString.fromUTF8(err).toErrorInstance(globalThis),
-                        };
-                        _ = listener.callWithGlobalThis(
-                            globalThis,
-                            &args,
-                        );
-                    }
-                }
+    }
+    pub fn emitError(this: *FSWatcher, err: string) void {
+        if (this.isClosed()) return;
+        defer this.close(true);
+
+        if (this.js_this != .zero) {
+            if (FSWatcher.listenerGetCached(this.js_this)) |listener| {
+                var args = [_]JSC.JSValue{
+                    JSC.ZigString.static("error").toValue(this.globalThis),
+                    JSC.ZigString.fromUTF8(err).toErrorInstance(this.globalThis),
+                };
+                _ = listener.callWithGlobalThis(
+                    this.globalThis,
+                    &args,
+                );
             }
         }
+    }
 
-        pub fn emit(this: *JSObject, file_name: string, comptime eventType: string) void {
-            if (this.globalThis) |globalThis| {
-                if (this.js_this != .zero) {
-                    if (JSObject.listenerGetCached(this.js_this)) |listener| {
-                        var filename: JSC.JSValue = JSC.JSValue.jsUndefined();
-                        if (file_name.len > 0) {
-                            if (this.encoding == .buffer)
-                                filename = JSC.ArrayBuffer.createBuffer(globalThis, file_name)
-                            else if (this.encoding == .utf8) {
-                                filename = JSC.ZigString.fromUTF8(file_name).toValueGC(globalThis);
-                            } else {
-                                // convert to desired encoding
-                                filename = Encoder.toStringAtRuntime(file_name.ptr, file_name.len, globalThis, this.encoding);
-                            }
-                        }
-                        var args = [_]JSC.JSValue{
-                            JSC.ZigString.static(eventType).toValue(globalThis),
-                            filename,
-                        };
-                        _ = listener.callWithGlobalThis(
-                            globalThis,
-                            &args,
-                        );
+    pub fn emit(this: *FSWatcher, file_name: string, comptime eventType: string) void {
+        if (this.js_this != .zero) {
+            if (FSWatcher.listenerGetCached(this.js_this)) |listener| {
```

Comment:
need to keep the value alive for the duration of this function call. It could finalize while the listener is called.